pyg_lib/csrc/ops/cpu/softmax_kernel.cpp

```diff
@@ -0,0 +1,258 @@
+#include <ATen/ATen.h>
+#include <ATen/Parallel.h>
+#include <torch/library.h>
+
+#include <algorithm>
+#include <cmath>
+#include <limits>
+#include <utility>
+#include <vector>
+
+namespace pyg {
+namespace ops {
+
+namespace {
+
+std::vector<int64_t> create_per_thread_groups(const int64_t* groups_ptr,
+                                              const int64_t n_groups,
+                                              const int64_t dim_size) {
+  std::vector<int64_t> new_groups = {0};
+  const auto avg_work_per_thread = at::divup(dim_size, at::get_num_threads());
+  int64_t cur_work = 0;
+  for (int64_t i = 0; i < n_groups; ++i) {
+    cur_work += groups_ptr[i + 1] - groups_ptr[i];
+    if (cur_work >= avg_work_per_thread) {
+      new_groups.push_back(i + 1);
+      cur_work = 0;
+    }
+  }
+  new_groups.push_back(n_groups);
+
+  return new_groups;
+}
+
+std::pair<std::vector<int64_t>, std::vector<int64_t>>
+precompute_data_access_patterns(const int64_t outer_range,
+                                const int64_t inner_range,
+                                const int64_t global_dim_size,
+                                const int64_t dim_stride) {
+  std::vector<int64_t> data_ids(outer_range * inner_range);
+  std::vector<int64_t> aux_ids(outer_range * inner_range);
+  for (int64_t i = 0; i < outer_range; ++i) {
+    const auto contiguous_offset = i * global_dim_size * dim_stride;
+    for (int64_t j = 0; j < inner_range; ++j) {
+      const auto k = i * inner_range + j;
+      const auto data_id = j + contiguous_offset;
+      const auto aux_id = k % dim_stride + (dim_stride * (k / inner_range));
+      data_ids[k] = data_id;
+      aux_ids[k] = aux_id;
+    }
+  }
+
+  return {std::move(data_ids), std::move(aux_ids)};
+}
+
+at::Tensor softmax_csr_forward_kernel_impl(const at::Tensor& src,
+                                           const at::Tensor& groups,
+                                           const int64_t dim) {
+  auto out = at::zeros_like(src);
+
+  AT_DISPATCH_FLOATING_TYPES(
+      src.scalar_type(), "softmax_csr_forward_kernel_impl", [&] {
+        const auto n_groups = groups.size(0) - 1;
+        const auto n_heads = src.numel() / src.size(dim);
```

Comment:
I assume that n_heads should be of integer type, so src.numel() will always be divisible by src.size(dim)? Maybe I'd worth to add assert about that. Same for the backward.

Reply:
I think it should be guaranteed. `numel()` can be represented as:
```python
numel_ = 1
for i in range(src.dim()):
    numel_ *= src.size(i)
```
As `numel` represents product of  `size`'s, it is divisible by them.

Reply:
Ok :)